### PR TITLE
add concept_difficulties to 20-hour course

### PR DESCRIPTION
**Pull request by Hamms**
Because the 20-hour course is not defined using `.level` files but rather a combination of `20-hour.script` and the apps levels, we need to do something a little special to add the concept difficulty tags to it.

Specifically, we need to give DSL-defined levels the ability to parse a `level_concept_difficulty` attribute, and then populate `20-hour.script` with the level_concept_difficulties.
### UPDATE

Turns out, a few levels in courses 1-3 are also hardcoded. This PR is therefore going to update those as well
# Methodology

level_concept_difficulties were downloaded from [the spreadsheet](https://docs.google.com/spreadsheets/d/1IUcvX50-olC2ud4gIaqYpKXwRnaI2bHFs1uJUA4HgrA/edit#gid=1637916396) and written into the file using this pretty hacky little script. Don't judge me.

``` ruby
require_relative('../config/environment')
require 'cdo/properties'

HEADERS_TO_FIX = {
  sequence_algorithms: :sequencing,
  repeat_until__while: :repeat_until_while,
  functions_wparams: :functions_with_params
}

def main(course)
  levels = {}
  CSV.foreach("./concept_tags/#{course}.csv", headers: true,  header_converters: :symbol) do |row|
    concept_difficulties = row.to_hash.compact.delete_if do |key, val|
      !LevelConceptDifficulty::CONCEPTS.include?(key.to_s)
    end

    HEADERS_TO_FIX.each do |key, val|
      concept_difficulties[val] = row[key] if row[key]
    end

    stage = row[:stage_position].to_i
    level = row[:level_position].to_i
    levels[stage] = {} unless levels.has_key? stage
    levels[stage][level] = {}
    concept_difficulties.each do |key, val|
      levels[stage][level][key.to_s] = val.to_i
    end
  end

  stage = 0
  level = 1
  IO.readlines("../config/scripts/#{course}.script").each do |line|
    if line.starts_with? "stage"
      stage = stage + 1
      level = 1
    end
    if line.starts_with? "level "
      if line.starts_with?("level 'blockly:") && levels[stage] && levels[stage][level]
        puts "level_concept_difficulty '#{JSON.generate(levels[stage][level])}'"
      end
      level = level + 1
    end
    puts line
  end
end

main(ARGV[0] || 'course1')
```

### CODE REVIEW

**Comment by Hamms** [comment]
FYI @ashercodeorg

File: dashboard/app/models/level.rb [hunk 1]
```diff
@@ -69,7 +69,8 @@ def serializable_hash(options=nil)
   # Rails won't natively assign one-to-one association attributes for
   # us, even though we've specified accepts_nested_attributes_for above.
   # So, we must do it manually.
-  def assign_attributes(attributes)
+  def assign_attributes(new_attributes)
+    attributes = new_attributes.stringify_keys
```
Comment (joshlory) [inline dashboard/app/models/level.rb:73]: What is the effect of this change?  IIRC there may be some fields that aren't stored as strings in .level files.

Comment (Hamms) [inline dashboard/app/models/level.rb:73]: It's simply [mirroring some functionality in `super`](https://github.com/rails/rails/blob/8b26431ba4cec4a5401dd1e5e50c15c11bec0703/activerecord/lib/active_record/attribute_assignment.rb#L29) to ensure that the next line will work the same whether this method was called with `'level_concept_difficulty'` or `:level_concept_difficulty` as an attribute.

**Comment by wjordan** [comment]
lgtm- don't have anything else to add other than a minor question: are there other scripts that use the  `blockly:level_num` syntax for referring to an apps-builtin level that we might also need to run this script against? Or are any other scripts using this syntax either obsolete, or not important enough to have concept difficulties attached to them?

**Comment by Hamms** [comment]
Oh, good catch! For some context, we currently only care about level concept difficulties for K-5 courses, since those are the only courses for which we've defined a tagging metric. This concept may someday be expanded to other courses, but that's not currently a concern.

However! Upon inspection, it does look like the other courses use a few `blockly:level_num`-style references, the 20-hour is just the only one that uses _exclusively_ those. I'll augment those courses with the appropriate data before merging this in.

Thanks!

**Pull request merged**